ioeffect/src/main/scala/scalaz/ioeffect/IOInstances.scala

```diff
@@ -46,14 +50,19 @@ private trait IOPlus[E] extends Plus[IO[E, ?]] {
   override def plus[A](a: IO[E, A], b: => IO[E, A]): IO[E, A] = a.catchAll(_ => b)
 }
 private class IOUnitMonadPlus extends IOMonadError[Unit] with IOPlus[Unit] with MonadPlus[IO[Unit, ?]] {
-  // note that if we were to implement MonadPlus for an arbitrary E, we would
-  // need to know the `zero` value (`Monoid`) and to be lawful we must consider
-  // if errors are empty (`Equal`) when implementing `plus` for two errors. This
-  // would therefore have a performance penalty on constructing the `Monad`, so
-  // we choose not to implement it.
   override def empty[A]: IO[Unit, A] = raiseError(())
 }
 
+private class IOMonadPlus[E: Monoid] extends IOMonadError[E] with MonadPlus[IO[E, ?]] {
+  override def plus[A](a: IO[E, A], b: => IO[E, A]): IO[E, A] =
+    a.catchAll { e1 =>
+      b.catchAll { e2 =>
+        IO.fail(Monoid[E].append(e1, e2))
+      }
```

Comment:
So if either fails, then it fails for that reason, if both fail, then it fails for a combination of reasons.

This is sensible and what I'd expect. 👍  on merging.